### PR TITLE
[Clone API]: Fix race between VM and restore readiness

### PR DESCRIPTION
**Pull request by iholder-redhat**
**What this PR does / why we need it**:
The new cloning API [was merged](https://github.com/kubevirt/kubevirt/pull/7336) [1] recently. Since then, we're seeing some CI failures (e.g. [this one](https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/directory/pull-kubevirt-e2e-k8s-1.23-sig-storage/1544549474116833280) [2]).

### The bug

The failures are coming from validating webhooks [3]. These webhooks don't allow VMs to start if a restore is still in progress.

The cloning API is making sure the restore object states that it's ready [4], but it doesn't check the VM's status.

Essentially, this is the current flow:
- Restore is complete. Its status is changed to reflect it.
- Then, these two things happen in parallel:
  - Cloning controller moves to "Succeeded" and considers the VM as ready
  - The VM updates its status to refrect it's ready

This race introduces flakieness, as in the test we:
- Wait for the clone to be finished
- Run the VM
- The VM is being rejected by webhooks since allegedly it isn't ready yet.

### The Fix

<del>In order to fix this, the cloning controller would wait for both the restore and the VM to be ready. This should resolve the flakieness.</del>

Following one of the [discussions in this PR](https://github.com/kubevirt/kubevirt/pull/8059#discussion_r917762046), we will solve this temporarily by waiting for `targetVm.Status.RestoreInProgress` to be nil in the clone controller functional tests.

The reason for this decision is that the clone controller should not depend on the target VM's status - but only on the restore object's status. To save CI lanes from being flaky we introduce a temporary dependency between the functional tests and the target VM's status. In the near future we will make sure the restore object is ready only when the target VM is ready.

---------

[1] https://github.com/kubevirt/kubevirt/pull/7336
[2] https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/directory/pull-kubevirt-e2e-k8s-1.23-sig-storage/1544549474116833280
[3] https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-api/webhooks/validating-webhook/admitters/vms-admitter.go#L533
[4] https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-controller/watch/clone/clone.go#L238

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/virt-controller/watch/clone/clone.go [hunk 1]
```diff
@@ -259,6 +260,24 @@ func (ctrl *VMCloneController) syncSourceVMTargetVM(key string, source *k6tv1.Vi
 		ctrl.logAndRecord(vmClone, TargetVMCreated, fmt.Sprintf("created target VM %s for clone %s", targetVMInfo.Name, vmClone.Name))
 		syncInfo.targetVMCreated = true
 
+		targetVM, ok := targetVMObj.(*k6tv1.VirtualMachine)
+		if !ok {
+			return addErrorToSyncInfo(syncInfo, fmt.Errorf("VM object from vm informer cannot be converted to *VirtualMachine: %v", err))
+		}
+
+		const vmNotReadyErrPattern = "target VM %s is still not ready - %s %s still in progress"
+		if targetVM.Status.SnapshotInProgress != nil {
```
Comment (ShellyKa13) [inline pkg/virt-controller/watch/clone/clone.go:269]: the snapshot for the clone is not on the target vm it was on the source vm, so seems redundant

File: pkg/virt-controller/watch/clone/clone.go [hunk 2]
```diff
@@ -346,10 +346,14 @@ func (ctrl *VMCloneController) updateStatus(origClone *clonev1alpha1.VirtualMach
...
+			assignPhase(clonev1alpha1.WaitingForTargetVMToBeReady)
+		}
+	}
+	if isInPhase(vmClone, clonev1alpha1.WaitingForTargetVMToBeReady) {
+		if syncInfo.targetVMReady {
 			vmClone.Status.SnapshotName = nil
 			vmClone.Status.RestoreName = nil
-			assignPhase(clonev1alpha1.Succeeded)
-
+			assignPhase(clonev1alpha1.WaitingForTargetVMToBeReady)
```
Comment (ShellyKa13) [inline pkg/virt-controller/watch/clone/clone.go:356]: this should have stayed succeeded isnt it?

File: pkg/virt-controller/watch/clone/clone.go [hunk 3]
```diff
@@ -259,6 +260,24 @@ func (ctrl *VMCloneController) syncSourceVMTargetVM(key string, source *k6tv1.Vi
...
+			logger.V(defaultVerbosityLevel).Infof(vmNotReadyErrPattern, targetVM.Name, "snapshot", *targetVM.Status.SnapshotInProgress)
+			return syncInfo
+		}
+		if targetVM.Status.RestoreInProgress != nil {
+			logger.V(defaultVerbosityLevel).Infof(vmNotReadyErrPattern, targetVM.Name, "restore", *targetVM.Status.RestoreInProgress)
+			return syncInfo
+		}
+
+		ctrl.logAndRecord(vmClone, TargetVMReady, fmt.Sprintf("target VM %s for clone %s is ready", targetVMInfo.Name, vmClone.Name))
+		syncInfo.targetVMCreated = true
```
Comment (ShellyKa13) [inline pkg/virt-controller/watch/clone/clone.go:279]: this should have been targetVMReady right?

File: pkg/virt-controller/watch/clone/clone.go (hunk 1 above)
Comment (iholder-redhat) [reply pkg/virt-controller/watch/clone/clone.go:269]: Yes, that's right. Thanks

File: pkg/virt-controller/watch/clone/clone.go (hunk 3 above)
Comment (iholder-redhat) [reply pkg/virt-controller/watch/clone/clone.go:279]: Absolutely

File: pkg/virt-controller/watch/clone/clone.go (hunk 2 above)
Comment (iholder-redhat) [reply pkg/virt-controller/watch/clone/clone.go:356]: :+1:

**Comment by iholder-redhat** [comment]
@ShellyKa13 Thanks a lot for your review!
Addressed your comments.

File: pkg/virt-controller/watch/clone/clone.go [hunk 4]
```diff
@@ -259,6 +260,19 @@ func (ctrl *VMCloneController) syncSourceVMTargetVM(key string, source *k6tv1.Vi
 		ctrl.logAndRecord(vmClone, TargetVMCreated, fmt.Sprintf("created target VM %s for clone %s", targetVMInfo.Name, vmClone.Name))
 		syncInfo.targetVMCreated = true
 
+		targetVM, ok := targetVMObj.(*k6tv1.VirtualMachine)
+		if !ok {
+			return addErrorToSyncInfo(syncInfo, fmt.Errorf("VM object from vm informer cannot be converted to *VirtualMachine: %v", err))
+		}
+
+		if targetVM.Status.RestoreInProgress != nil {
+			logger.V(defaultVerbosityLevel).Infof("target VM %s is still not ready - restore %s still in progress", targetVM.Name, *targetVM.Status.RestoreInProgress)
```
Comment (ShellyKa13) [inline pkg/virt-controller/watch/clone/clone.go:269]: small thing is that in a previous log you write restore is ready to use and then here it will be a bit confusing that its written still in progress. maybe rephrase to status of restore not updated yet on the VM or something like that.

File: pkg/virt-controller/watch/clone/clone_base.go [hunk 5]
```diff
@@ -30,6 +30,7 @@ const (
 	RestoreCreated  Event = "RestoreCreated"
 	RestoreReady    Event = "RestoreReady"
 	TargetVMCreated Event = "TargetVMCreated"
+	TargetVMReady   Event = "TargetVMCreated"
```
Comment (ShellyKa13) [inline pkg/virt-controller/watch/clone/clone_base.go:33]: the msg of the event should be Ready not Created

File: pkg/virt-controller/watch/clone/clone.go (hunk 4 above)
Comment (iholder-redhat) [reply pkg/virt-controller/watch/clone/clone.go]: done

File: pkg/virt-controller/watch/clone/clone_base.go (hunk 5 above)
Comment (iholder-redhat) [reply pkg/virt-controller/watch/clone/clone_base.go]: done

File: pkg/virt-controller/watch/clone/clone.go [hunk 6]
```diff
@@ -246,10 +247,10 @@ func (ctrl *VMCloneController) syncSourceVMTargetVM(key string, source *k6tv1.Vi
```
[the remark this answers is not in the document]
Comment (iholder-redhat) [reply pkg/virt-controller/watch/clone/clone.go:247]: I'm not sure about that. Some environments don't change the verbosity level and then it's very hard to debug without any log prints.

When the tests aren't flakie anymore this shouldn't spam the logs. Let's hope that's the case after this PR :)

**Comment by iholder-redhat** [comment]
@ShellyKa13 thanks again! comments addressed.

**Comment by brybacki** [comment]
There is actually one thing I do not like here, the commit messages. 

You have awesome and detailed description here, even the title on PR is good. But there is nothing like "In order to fix this, the cloning controller would wait for both the restore and the VM to be ready. This should resolve the flakieness." mentioned in the commit messages. I think the commit should at least mention what or why was fixed.

File: pkg/virt-controller/watch/clone/clone.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (ShellyKa13) [reply pkg/virt-controller/watch/clone/clone.go:247]: I dont think its related to this PR really, since its in a different phase of the clone, this print will still be printed until the restore is done, and since there is a print after it if you dont see it you can pretty much understand where you stand. but ok we can decide about it in the future

File: pkg/virt-controller/watch/clone/clone.go [hunk 7]
```diff
@@ -259,6 +260,19 @@ func (ctrl *VMCloneController) syncSourceVMTargetVM(key string, source *k6tv1.Vi
 		ctrl.logAndRecord(vmClone, TargetVMCreated, fmt.Sprintf("created target VM %s for clone %s", targetVMInfo.Name, vmClone.Name))
 		syncInfo.targetVMCreated = true
 
+		targetVM, ok := targetVMObj.(*k6tv1.VirtualMachine)
+		if !ok {
+			return addErrorToSyncInfo(syncInfo, fmt.Errorf("VM object from vm informer cannot be converted to *VirtualMachine: %v", err))
+		}
+
+		if targetVM.Status.RestoreInProgress != nil {
+			logger.V(defaultVerbosityLevel).Infof("status of restore %s still not updated in VM %s statue", *targetVM.Status.RestoreInProgress, targetVM.Name)
```
Comment (ShellyKa13) [inline pkg/virt-controller/watch/clone/clone.go:269]: small typo statue -> status

Comment (iholder-redhat) [reply pkg/virt-controller/watch/clone/clone.go]: done, thanks!

**Review by ShellyKa13 — COMMENTED** [review]
/lgtm

**Comment by brybacki** [comment]
The tests here fail on VirtualMachineClone, there is still some problem

File: pkg/virt-controller/watch/clone/clone.go [hunk 8]
```diff
@@ -178,7 +179,7 @@ func (ctrl *VMCloneController) syncSourceVMTargetVM(key string, source *k6tv1.Vi
 		snapshot = obj.(*snapshotv1alpha1.VirtualMachineSnapshot)
 		logger.Infof("found snapshot %s for clone %s", snapshot.Name, vmClone.Name)
 
-		if snapshot.Status.ReadyToUse == nil || *snapshot.Status.ReadyToUse == false {
+		if snapshot.Status == nil || snapshot.Status.ReadyToUse == nil || *snapshot.Status.ReadyToUse == false {
```
Comment (ShellyKa13) [inline pkg/virt-controller/watch/clone/clone.go:182]: maybe use existing function of checking if snapshot is ready? vmSnapshotReady
and also for restore you have function -> vmRestoreProgressing to use in line 239

Comment (iholder-redhat) [reply pkg/virt-controller/watch/clone/clone.go]: Done, thanks!

File: pkg/virt-controller/watch/clone/clone.go [hunk 9]
```diff
@@ -259,6 +262,18 @@ func (ctrl *VMCloneController) syncSourceVMTargetVM(key string, source *k6tv1.Vi
 		ctrl.logAndRecord(vmClone, TargetVMCreated, fmt.Sprintf("created target VM %s for clone %s", targetVMInfo.Name, vmClone.Name))
 		syncInfo.targetVMCreated = true
 
+		targetVM, ok := targetVMObj.(*k6tv1.VirtualMachine)
+		if !ok {
+			return addErrorToSyncInfo(syncInfo, fmt.Errorf("VM object from vm informer cannot be converted to *VirtualMachine: %v", err))
+		}
+
+		if targetVM.Status.RestoreInProgress != nil {
+			return addErrorToSyncInfo(syncInfo, fmt.Errorf("status of restore %s still not updated in VM %s status", *targetVM.Status.RestoreInProgress, targetVM.Name))
```
Comment (rmohr) [inline pkg/virt-controller/watch/clone/clone.go:271]: Do I read this right that we are reporting a real error to the controller loop? I am not sure this is an error, that seems to be like a normal we-have-to-wait-until-it-is-done check. In this case we normally don't return errors (printing a log message for debug purposes is of course fine). We would even continue updating the status section if not in sync, but we would not return with an error, which would lead to a back-off retry. Not 100% sure this is what is happening here ,but it looks like it.

Comment (iholder-redhat) [reply pkg/virt-controller/watch/clone/clone.go:271]: Yes, you're reading it correctly.

The reason I chose this path is since I didn't want to pass a VM informer and add an event handler to it. With the current implementation, the clone controller does not track VMs. My concern was that it would introduce an overhead to track all VMs, wake up for every one of their changes, check if the change is relevant (99% it's not), then go back to sleep.

It's important to note that this is now being added since there's a small delay between the restore reflecting that it's ready and the VM reflecting that the restore is ready. This delay is expected to be very small, so I thought returning an error and trying again is reasonable, perhaps more reasonable than waking up for every VM change with a very very low probability of the change being relevant.

p.s. in the original clone implementation, the `SyncInfo` struct had information on requeuing. Perhaps I can do the same now - simply requeue (after, say, 1 second) in this scenario. This would require some more changes, which again I'm not sure are worth the trouble for this corner case, but it's an option.

@mhenriks @rmohr WDYT?

Comment (rmohr) [reply pkg/virt-controller/watch/clone/clone.go:271]: Probably best to wait for @mhenriks. I was not part of the design decisions on this one, but I would be cautious. This really reports an error, like including it in our controller-prometheus metrics, making it hard to see real errors. And if I understand you correct, we just talk about an eventually-consistency-race which is pretty normal, and not a "faulty" race, right?

Comment (iholder-redhat) [reply pkg/virt-controller/watch/clone/clone.go:271]: Yes, you got it right. Thanks for raising this concern.
Alright, let's wait for @mhenriks to bring his thoughts

Comment (mhenriks) [reply pkg/virt-controller/watch/clone/clone.go:271]: My preference would be that the clone layer does not have to know/care about `targetVM.Status.RestoreInProgress`.  I think it would be better if the restore is not complete until `targetVM.Status.RestoreInProgress` is nil.  But that will take a bit of work in the restore controller.

For now, maybe the best path is to simply have the test wait for `RestoreInProgress` to be nil before starting the VM?  Then we can refactor the restore controller at our convenience?

Comment (iholder-redhat) [reply pkg/virt-controller/watch/clone/clone.go:271]: > My preference would be that the clone layer does not have to know/care about `targetVM.Status.RestoreInProgress`. I think it would be better if the restore is not complete until `targetVM.Status.RestoreInProgress` is nil. But that will take a bit of work in the restore controller.
> 
> For now, maybe the best path is to simply have the test wait for `RestoreInProgress` to be nil before starting the VM? Then we can refactor the restore controller at our convenience?

Sounds great :+1: 
Let's try that.

**Comment by rmohr** [comment]
/lgtm
/approve

/hold

@iholder-redhat I don't think that the release note is not correct now. We may not need one.

Giving @mhenriks a chance to have a final look.

**Comment by iholder-redhat** [comment]
> @iholder-redhat I don't think that the release note is not correct now. We may not need one.

Updated, thanks for catching that :+1:

**Comment by iholder-redhat** [comment]
CI failures aren't related to clone
/retest-required

**Comment by ShellyKa13** [comment]
@iholder-redhat I lgtm before I read your message. You say that 30 seconds is not enough to wait  for the RestoreInProgress to be removed from the vm and make it ready to be started, but it doesnt make sense to me, espcially that all the restore tests actually dont wait for this and they just start the vm after the restore and nothing fails, so Im starting to be a little skeptic that this is the real issue here. What do you think?

**Comment by ShellyKa13** [comment]
@iholder-redhat 
Maybe I wasn't so clear with what I meant.. I wasn't saying that the issue is not that the RestoreInProgress is not on the vm preventing it from starting - we saw it together in the logs..
But it is strange to me that it takes so much time for it to be updated on the vm, maybe it's ok, I'm not sure, but as mentioned, in  the restore tests we do start the restored vm right after the restore is finished whether its the same vm or a new target vm that you introduced, so it doesnt seem like the issue is restoring to a new target vm..
My point is, it's possible that increasing the time to 90 seconds will cause the tests to pass but my question is: are we fine with that? there might be something slowing this update from some reason that we might want to look into? I don't know.. I was just wondering, I'll let @mhenriks give the final say :)

**Comment by ShellyKa13** [comment]
@iholder-redhat also, if this fixes the issue for now, I do think it's better to get this PR merged and if needed look into what I raised later because it fails a lot of other PRs

**Comment by brianmcarey** [comment]
/hold 

Holding merge queue PRs due to flakes introduced to the sig-operator lanes.

**Comment by dhiller** [comment]
We have restarted the operator lanes also to see them succeed after we made the virt-exportserver quay.io repo public.

Even though we still see failures on the storage lanes, we think this PR will improve the situation regarding clone flakes. 

/override pull-kubevirt-e2e-k8s-1.23-sig-storage-nonroot

We will look at the flakefinder report on Monday and then decide to quarantine the clone tests when we have more data.
